### ISSUE TITLE
feat: support for custom url encoding

### ISSUE THREAD

**Issue by JesusTheHun** [trpc/trpc#3590]
### Describe the feature you'd like to request

Right now tRPC only support data transformers for the body of requests.
When performing a query, the query string can be quite large. Some providers like `some-that-removed-their-free-plan-recently-so-I-wont-advertise-their-name` limit the URL to 8KB. While this is substantial and meet most of our needs, sometimes it's not enough.

I would like the ability to pass my own url serializer. Some [examples](https://tanstack.com/router/v1/docs/guide/custom-search-param-serialization) are given in the `TanStack Router` documentation.

<details>
<summary>tRPC current getUrl()</summary>

```ts
// https://github.com/trpc/trpc/blob/v10.8.1/packages/client/src/links/internals/httpUtils.ts#L96

export function getUrl(opts: HTTPRequestOptions) {
  let url = opts.url + '/' + opts.path;
  const queryParts: string[] = [];
  if ('inputs' in opts) {
    queryParts.push('batch=1');
  }
  if (opts.type === 'query') {
    const input = getInput(opts);
    if (input !== undefined) {
      queryParts.push(`input=${encodeURIComponent(JSON.stringify(input))}`);
    }
  }
  if (queryParts.length) {
    url += '?' + queryParts.join('&');
  }
  return url;
}
```
</details>

### Describe the solution you'd like to see

Configuring the client with a custom query parameter seems to be a perfectly viable solution to me.

### Describe alternate solutions

I haven't found a workaround so far. I thought about using a `mutation` instead, so it would be a post, but then I would not be able to use `useInfiniteQueries`. Dealbreaker.

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR implementing this feature!

<sub>[TRP-45](https://linear.app/trpc/issue/TRP-45/feat-support-for-custom-url-encoding)</sub>

## Funding
* You can sponsor this specific effort via a [Polar.sh](https://polar.sh) pledge below
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/trpc/trpc/issues/3590">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/trpc/trpc/issues/3590/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/trpc/trpc/issues/3590/pledge.svg">
</picture>
</a>

**Comment by KATT** [comment]
https://trpc.io/docs/links/httpBatchLink#setting-a-maximum-url-length

**Comment by JesusTheHun** [comment]
I've seen that option, sadly the payload is for a single request.

**Comment by KATT** [comment]
THis use case is better solved by #3910

**Comment by KATT** [comment]
Will close this as we're now implementing #3910, reopen the feature request if you think it's still needed
